src/model.jl

```diff
@@ -66,21 +66,68 @@ function _make_bus_demand_flexibility_amount(
+    
+    # DOE flexibility, convert to MW, DOE timestamps already aligned
+    if demand_flexibility.enable_doe_flexibility
+        doe_flex_pct = Matrix(
+            demand_flexibility.doe_flex_amt[start_index:end_index, 2:end]
+        )
+        doe_flex_mw = doe_flex_pct .*= permutedims(bus_demand[sets.doe_flexible_bus_idx, :])
+        for i in 1:length(sets.doe_flexible_bus_idx)
+            simulation_demand_flex_amt_up[:, sets.doe_flexible_bus_idx[i]] = doe_flex_mw[:, i]
+            simulation_demand_flex_amt_dn[:, sets.doe_flexible_bus_idx[i]] = doe_flex_mw[:, i]
+        end
+    end 
+    
+    # replace columns with user input numbers
+    csv_flex_amt_up = Matrix(
         demand_flexibility.flex_amt_up[start_index:end_index, 2:end]
     )
-    # Demand flexibility down profile
-    simulation_demand_flex_amt_dn = Matrix(
+    csv_flex_amt_dn = Matrix(
         demand_flexibility.flex_amt_dn[start_index:end_index, 2:end]
     )
+    
+    csv_flex_cost_up = Matrix(
+        demand_flexibility.cost_up[:, 2:end]
+    )
+    csv_flex_cost_dn = Matrix(
+        demand_flexibility.cost_dn[:, 2:end]
+    )
 
-    # percentage flexibility from input
-    if demand_flexibility.enable_doe_flexibility
-        simulation_demand_flex_amt_up .*= permutedims(bus_demand[sets.flexible_bus_idx, :])
-        simulation_demand_flex_amt_dn .*= permutedims(bus_demand[sets.flexible_bus_idx, :])
+    for i in 1:length(sets.csv_flexible_bus_idx)
+        simulation_demand_flex_amt_up[:, sets.csv_flexible_bus_idx[i]] = csv_flex_amt_up[:, i]
+        simulation_demand_flex_amt_dn[:, sets.csv_flexible_bus_idx[i]] = csv_flex_amt_dn[:, i]
+        
+        new_demand_flex_cost_up[:, sets.csv_flexible_bus_idx[i]] = csv_flex_cost_up[:, i]
+        new_demand_flex_cost_dn[:, sets.csv_flexible_bus_idx[i]] = csv_flex_cost_dn[:, i]
     end
 
+    # remove non-flexible columns
+    simulation_demand_flex_amt_up = simulation_demand_flex_amt_up[:, sets.flexible_bus_idx]
+    simulation_demand_flex_amt_dn = simulation_demand_flex_amt_dn[:, sets.flexible_bus_idx]
+    new_demand_flex_cost_up = new_demand_flex_cost_up[:, sets.flexible_bus_idx]
+    new_demand_flex_cost_dn = new_demand_flex_cost_dn[:, sets.flexible_bus_idx]
+    
+    # construct new cost Dataframe with appropriate dimension
+    demand_flexibility.cost_up = DataFrames.DataFrame(new_demand_flex_cost_up, :auto)
+    demand_flexibility.cost_dn = DataFrames.DataFrame(new_demand_flex_cost_dn, :auto)
+    
+    DataFrames.rename!(demand_flexibility.cost_dn, Symbol.(sets.flexible_bus_idx))
+    DataFrames.rename!(demand_flexibility.cost_dn, Symbol.(sets.flexible_bus_idx))
+    DataFrames.insertcols!(
+        demand_flexibility.cost_up, 1, :"UTC Time" => demand_flexibility.flex_amt_up[!, "UTC Time"]
+    )
+    DataFrames.insertcols!(
+        demand_flexibility.cost_dn, 1, :"UTC Time" => demand_flexibility.flex_amt_up[!, "UTC Time"]
+    )    
+    
```

Comment:
Maybe we can change this function (`_make_bus_demand_flexibility_amount`) to also output the cost profiles? We could output the pared-down profiles that are later calculated [here](https://github.com/Breakthrough-Energy/REISE.jl/blob/develop/src/model.jl#L551) and [here](https://github.com/Breakthrough-Energy/REISE.jl/blob/develop/src/model.jl#L561), which would match the similarly pared-down profiles that are already returned by this function. Instead of updating the DataFrames directly in the `DemandFlexibility` object, we could just create local variables for use in this function. Doing this would allow us to keep the structs as immutable (which is quite a bit faster than if the structs are mutable, as far as I understand).